main/FatesInventoryInitMod.F90

```diff
@@ -374,6 +377,82 @@ subroutine initialize_sites_by_inventory(nsites,sites,bc_in)
+            ! first generate a new head of the linked list.
+            head_of_unsorted_patch_list => sites(s)%youngest_patch%older
+
+            ! reset the site-level patch linked list, keeping only the youngest patch.
+            sites(s)%youngest_patch%older => null()
+            sites(s)%youngest_patch%younger => null()
+            sites(s)%oldest_patch   => sites(s)%youngest_patch
+
+            ! loop through each patch in the unsorted LL, peel it off,
+            ! and insert it into the new, sorted LL
+            do while(associated(head_of_unsorted_patch_list))
+
+               ! first keep track of the next patch in the old (unsorted) linked list
+               next_in_unsorted_patch_list => head_of_unsorted_patch_list%older
+               
+               ! check the two end-cases
+
+               ! Youngest Patch
+               if(sum(head_of_unsorted_patch_list%pft_agb_profile(:,:)) <= &
+                    sum(sites(s)%youngest_patch%pft_agb_profile(:,:)))then
+                  head_of_unsorted_patch_list%older                  => sites(s)%youngest_patch
+                  head_of_unsorted_patch_list%younger                => null()
+                  sites(s)%youngest_patch%younger => head_of_unsorted_patch_list
+                  sites(s)%youngest_patch         => head_of_unsorted_patch_list
+
+                  ! Oldest Patch
+               else if(sum(head_of_unsorted_patch_list%pft_agb_profile(:,:)) > &
+                    sum(sites(s)%oldest_patch%pft_agb_profile(:,:)))then
+                  head_of_unsorted_patch_list%older              => null()
+                  head_of_unsorted_patch_list%younger            => sites(s)%oldest_patch
+                  sites(s)%oldest_patch%older => head_of_unsorted_patch_list
+                  sites(s)%oldest_patch       => head_of_unsorted_patch_list
+
+                  ! Somewhere in the middle
+               else
+                  currentpatch => sites(s)%youngest_patch
+                  do while(associated(currentpatch))
+                     olderpatch => currentpatch%older
+                     if(associated(currentpatch%older)) then
+                        if(sum(head_of_unsorted_patch_list%pft_agb_profile(:,:)) >= &
+                             sum(currentpatch%pft_agb_profile(:,:)) .and. &
+                             sum(head_of_unsorted_patch_list%pft_agb_profile(:,:)) < &
+                             sum(olderpatch%pft_agb_profile(:,:))) then
+                           ! Set the new patches pointers
+                           head_of_unsorted_patch_list%older   => currentpatch%older
+                           head_of_unsorted_patch_list%younger => currentpatch
+                           ! Fix the patch's older pointer
+                           currentpatch%older => head_of_unsorted_patch_list
+                           ! Fix the older patch's younger pointer
+                           olderpatch%younger => head_of_unsorted_patch_list
```

Comment:
@ckoven do you think it's worth it to insert a `exit` statement in this inner loop to speed up the sorting?  I'm thinking of a situation in which you've moved head, but then current/older keeps itereting down the list via this `do while(associated(currentpatch))` loop.  There might come a later current/older pair that that head could fit between which would push the head further down the list, correct?

Reply:
yeah, that makes sense.

Reply:
Adding the `exit` cut the sort time in about half.  Without `exit`:
```
 BEGIN PATCH LOOP
 Number of patches:        1251
 Total Shifts of Head:      775976
```
Here 'Head' is `head_of_unsorted_patch_list`.  This output is here:
```
/home/glemieux/scratch/clmed-cases/ckoven-patch-sort-test-bci-5yr-iter.C137514abc-Fc9f25c1c.2019-10-01/run/logs/lnd.log.191001-165731
```

With `exit`:
```
 BEGIN PATCH LOOP
 Number of patches:        1251
 Total Shifts of Head:      406927
```
```
/home/glemieux/scratch/clmed-cases/ckoven-patch-sort-test-bci-5yr-iter-exit.C137514abc-Fc9f25c1c.2019-10-01/run/logs/lnd.log.191001-223448
```

Ran a compare test suite on Cheyenne on my branch and all expected passed.  Output is here:

```
/glade/u/home/glemieux/scratch/clmed-tests/ckoven-sort-patches-nearzero.fates.cheyenne.intel.C9c12e402-F7bbcd3ca
```